### DISCUSSION TITLE
Postprocessor for verifying mass flow rate conservation

### DISCUSSION THREAD

**Discussion by js-jixu** [idaholab/moose#22307] [Q&A General] [answered]
Hi, experts.

I have a model of fluid-solid conjugate heat transfer. Now I want to verify that the `mass flow rate` is conserved within the fluid region. That is, whether the mass flow rate at the inlet is equal to the mass flow rate at the outlet. Is there any way in MOOSE to do this?

I know that `VolumetricFlowRate` might be able to do it. But the `advised_mat_prop` parameter in `VolumetricFlowRate` requires that the material property be `Functor`, but my material type is `ADParsedMaterial`, so `VolumetricFlowRate` is not suitable.

Actually I've used the following syntax to do a simple calculation:

```
[AuxKernels]
  [./speed_auxk]
    type = VectorVariableMagnitudeAux
    variable = speed
    vector_variable = velocity
    block = 'fluid'
  [../]
[]

[Materials]
  [./rho_fluid-nonAD]
    type = ParsedMaterial
    f_name = rho_fluid_noAD
    function = '(11096 - 1.3236 * Tf) * 1e-6'
    args = 'Tf'
    block = 'fluid'
  [../]
  [./mass_flow_rate_density_mat]
    type = ParsedMaterial
    block = 'fluid'
    f_name = mass_flow_rate_density
    args = speed
    material_property_names = rho_fluid_noAD
    function = 'rho_fluid_noAD * speed'
  [../]
[]

[Postprocessors]
  [./outlet_mass_flow_rate]
    type = SideIntegralMaterialProperty
    boundary = 'fluid_top'
    property = mass_flow_rate_density
  [../]
  [./inlet_mass_flow_rate]
    type = SideIntegralMaterialProperty
    boundary = 'fluid_bottom'
    property = mass_flow_rate_density
  [../]
[]
```

However, the relative error of `outlet_mass_flow_rate` and `inlet_mass_flow_rate` is around one thousandth, which is obviously not a satisfactory accuracy. So I want to ask if there is any Postprocessor that can calculate mass flow rate directly?

**Comment by GiudGiud** [comment, marked as answer]
Hello

This is the same issue as in the previous discussion around energy conservation
https://github.com/idaholab/moose/issues/22074

Side integrals are not appropriate to look at conservation with FE. You will need to make a mass conservation version of the [INSElementIntegralEnergyAdvection.C](https://github.com/idaholab/moose/pull/22075/files#diff-b03cb83497d9b5bd4c66da77588a01c784c90dbcc392e2bed4ab4730948cfb7f)

But really if energy is conserved you can trust that mass is conserved as well

Guillaume

**Reply by js-jixu** [reply]
I'm doing thermomechanical coupling, and the expansion of the solid will cause the outlet channel to narrow and the outlet velocity to increase. So I need to verify that the outlet mass flow rate is the same as the inlet mass flow rate. 

> But really if energy is conserved you can trust that mass is conserved as well

If I use INSElementIntegralEnergyAdvvection to verify the energy conservation of thermomechanical coupling, does it mean that mass is naturally conserved？ Then the mass flow rate is also conserved.

**Reply by GiudGiud** [reply]
It would be hard to conserve energy if mass was not conserved.
If energy is not conserved, then I would investigate if mass is conserved

**Reply by hugary1995** [reply]
You probably want to review thermodynamics. For incompressible flow the mass balance reduces to div(u) = 0, where u is the fluid velocity.

**Reply by js-jixu** [reply]
Thanks for your advice @hugary1995 @GiudGiud . I've used `[INSElementIntegralEnergyAdvection]` to verifiy energy conservation, which relative error is 3e-10. Does it  mean that mass is naturally conserved?

> For incompressible flow the mass balance reduces to div(u) = 0, where u is the fluid velocity.

And how can I verify whether div(u) equals to 0?

**Reply by GiudGiud** [reply]
It means energy is conserved. As I mentioned it would be hard not conserving mass, while conserving energy. However, I'm sure you could engineer a solution which does one without the other.
If you care to check, then you need to create the object I mentioned

There is a divergence auxkernel you can use, though it wont account for stabilization terms.

**Reply by hugary1995** [reply]
So if say it without jargon: You don't check mass conservation for INS... You are not solving for density.

**Reply by js-jixu** [reply]
It looks like creating the object is necessary to solve this problem. I will give it a try.👀

**Reply by hugary1995** [reply]
:-/

Again, feel free to go over the derivation of navier stokes equations yourself. Mass balance is an _assumption_ not a consequence.

**Reply by GiudGiud** [reply]
You are solving that equation for conservation of mass. If you dont converge it much, you can see mass not being converged because velocities entering the cell are larger than exiting the cell (for finite volume, for FE it s more a volumetric problem). Mass balance is the residual of mass equation pretty much

With the time-dependent modification of the meshes, I can understand their desire to check mass conservation.

**Reply by js-jixu** [reply]
I'm a little confused and dizzy now... What I can understand now is that if I want to verify that the mass flow rate is conserved, in another words, the inlet mass flow rate and the outlet mass flow rate are equal, I need to create an object like `INSElementIntegralEnergyAdvection` . And this object needs to take density and speed into consideration. The product of density and velocity is then integrated over the boundary. If the integral at the inlet boundary is equal to the integral at the outlet boundary, then it is proved that the mass flow rate is conserved. Is there something wrong with my understanding?

**Reply by GiudGiud** [reply]
Advection in a finite element discretization is a volumetric term, so you can only look at volumetric balances, not surface balances. You integrate the advection term contribution over a group of elements (from one to all of the mesh), and that lets you know what the contribution of advection is to the balance in the domain. In the absence of volumetric sources or sinks , we expect 0.

Or at least that's my understanding!

**Reply by hugary1995** [reply]
1. For any fluid, balance of mass is `dot(rho) + div(rho u) = 0`
2. For incompressible fluid, 1 reduces to `div(u) = 0`
3. @js-jixu the method you proposed above to integrate `rho u` along boundary is equivalent to measuring `div(u)` over the volume, owing to divergence theorem, albeit stabilization.
4. The `div(u)` term is exactly one of the equations you solve using the navier stokes module, e.g. `INSADMass` and `INSADMassPSPG` or their FV counterparts.
5. Due to 3 and 4, you are effectively looking at the residual norm of your discrete system of equations, which is guaranteed to satisfy the tolerances you set upon convergence.
6. @GiudGiud has a good point about your specific problem where you are deforming the mesh and solving on the displaced mesh. Although I don't think that'll change any of the above points, it's probably not a bad idea to triple check...

Let me know if any of the above items isn't clear. I'm happy to discuss.

**Reply by hugary1995** [reply]
> Advection in a finite element discretization is a volumetric term, so you can only look at volumetric balances, not surface balances. You integrate the advection term contribution over a group of elements (from one to all of the mesh), and that lets you know what the contribution of advection is to the balance in the domain. In the absence of volumetric sources or sinks , we expect 0.
> 
> Or at least that's my understanding!

That's correct because we don't apply divergence theorem on the advection term. We are going back to one of our previous discussions lol

**Reply by js-jixu** [reply]
These discussions helped me sort out my ambiguity and I can understand it now (one of the reasons I was confused before was that I forgot the divergence theorem) !!! Thanks a lot for your patience @hugary1995 @GiudGiud (❁´◡`❁).
